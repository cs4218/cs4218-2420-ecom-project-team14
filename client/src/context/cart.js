import axios from "axios";
import { createContext, useContext, useEffect, useReducer } from "react";
import toast from "react-hot-toast";
import { useAuth } from "./auth";

const CartContext = createContext();

const cartReducer = (state, action) => {
  switch (action.type) {
    case "ADD_TO_CART": {
      const { slug, price, productId } = action.payload;
      toast.success("Add to Cart Successfully");
      return {
        ...state,
        [slug]: state[slug]
          ? { quantity: state[slug].quantity + 1, price, productId}
          : { quantity: 1, price, productId},
      };
    }
    case "REMOVE_FROM_CART": {
      const newState = { ...state };
      delete newState[action.payload.slug];
      toast.success("Remove from Cart Successfully");
      return newState;
    }
    case "UPDATE_QUANTITY": {
      const { slug, quantity, price, productId } = action.payload;
      if (quantity <= 0) {
        const newState = { ...state };
        delete newState[slug];
        return newState;
      }
      toast.success("Update Cart Quantity Successfully");
      return {
        ...state,
        [slug]: { ...state[slug], quantity, price, productId },
      };
    }
    case "CLEAR_CART":
      toast.success("Cart Cleared Successfully");
      return {};
    default:
      return state;
  }
};

export const CartProvider = ({ children }) => {
  const auth = useAuth();

  const user = auth.user ?? "guest";

  const [cart, dispatch] = useReducer(cartReducer, {}, () => {
    const localData = localStorage.getItem(`cart-${user}`);
    return localData ? JSON.parse(localData) : {};
  });

  useEffect(() => {
    localStorage.setItem(`cart-${user}`, JSON.stringify(cart));
  }, [cart]);

  const addToCart = async (slug) => {
    const { data } = await axios.get(`/api/v1/product/get-product/${slug}`);
    if (!data.product) {
      toast.error("Item does not exist");
      return;
    }

    const inventory = data.product.quantity;

    if (cart[slug] && cart[slug].quantity + 1 > inventory) {
      toast.error("Error added to cart: Not enough inventory");
      return;
    }

    if (!data.product.price) {
      toast.error("Error added to cart: Price of product not available");
      return;
    }

    dispatch({ type: "ADD_TO_CART", payload: { slug, price: data.product.price, productId: data.product._id } });
  };

  const removeFromCart = (slug) => {
    dispatch({ type: "REMOVE_FROM_CART", payload: { slug } });
  };

  const updateQuantity = async (slug, quantity) => {
    const { data } = await axios.get(`/api/v1/product/get-product/${slug}`);
    if (!data.product) {
      toast.error("Item does not exist");
      return;
    }

    const inventory = data.product.quantity;

    if (cart[slug] && quantity > inventory) {
      toast.error("Error updating quantity: Not enough inventory");
      return;
    }

    if (!data.product.price) {
      toast.error("Error added to cart: Price of product not available");
      return;
    }

    dispatch({ type: "UPDATE_QUANTITY", payload: { slug, quantity, price: data.product.price, productId: data.product._id } });
  };

  const clearCart = () => {
    dispatch({ type: "CLEAR_CART" });
  };

  return (
    <CartContext.Provider
      value={{
        cart,
        addToCart,
        removeFromCart,
        updateQuantity,
        clearCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export const useCart = () => {
  return useContext(CartContext);
};
