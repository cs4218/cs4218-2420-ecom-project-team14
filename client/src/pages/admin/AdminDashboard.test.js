import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import AdminDashboard from "./AdminDashboard";
import { useAuth } from "../../context/auth";

// Mock all dependencies
jest.mock("../../components/Layout", () => ({ children }) => (
  <div data-testid="mock-layout">{children}</div>
));

jest.mock("../../components/AdminMenu", () => () => (
  <div data-testid="mock-admin-menu">Admin Menu</div>
));

jest.mock("../../context/auth", () => ({
  useAuth: jest.fn(),
}));

describe("AdminDashboard Component", () => {
  const mockAuthData = {
    user: {
      name: "Nadya Yuki",
      email: "yukiwuki@gmail.com",
      phone: "12345678",
    },
  };

  beforeEach(() => {
    useAuth.mockReturnValue([mockAuthData]);
  });

  // Rendering Tests
  describe("Component Rendering", () => {
    it("renders AdminDashboard without crashing", () => {
      render(<AdminDashboard />);
      expect(screen.getByTestId("mock-layout")).toBeInTheDocument();
    });

    it("renders AdminMenu inside AdminDashboard", () => {
      render(<AdminDashboard />);
      expect(screen.getByTestId("mock-admin-menu")).toBeInTheDocument();
    });
  });

  // Structure Test
  describe("Component Structure", () => {
    it("contains correct Bootstrap grid structure", () => {
      render(<AdminDashboard />);
      const container = screen.getByTestId("mock-layout").firstChild;
      expect(container).toBeInTheDocument();
      expect(container).toHaveClass("container-fluid", "m-3", "p-3");

      const columns = container.querySelectorAll(".col-md-3, .col-md-9");
      expect(columns).toHaveLength(2);
      expect(columns[0]).toHaveClass("col-md-3");
      expect(columns[1]).toHaveClass("col-md-9");
    });
  });

  // Content Test
  describe("Content", () => {
    it("displays admin name, email, and phone number", () => {
      render(<AdminDashboard />);
      expect(screen.getByText(`Admin Name : ${mockAuthData.user.name}`)).toBeInTheDocument();
      expect(screen.getByText(`Admin Email : ${mockAuthData.user.email}`)).toBeInTheDocument();
      expect(screen.getByText(`Admin Contact : ${mockAuthData.user.phone}`)).toBeInTheDocument();
    });
  });

  // Snapshot Test
  describe("Snapshot", () => {
    it("matches snapshot", () => {
      const { container } = render(<AdminDashboard />);
      expect(container).toMatchSnapshot();
    });
  });
});
