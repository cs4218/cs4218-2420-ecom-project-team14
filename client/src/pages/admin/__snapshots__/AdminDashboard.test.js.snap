// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminDashboard Component Snapshot matches snapshot 1`] = `
<div>
  <div
    data-testid="mock-layout"
  >
    <div
      class="container-fluid m-3 p-3"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          <div
            data-testid="mock-admin-menu"
          >
            Admin Menu
          </div>
        </div>
        <div
          class="col-md-9"
        >
          <div
            class="card w-75 p-3"
          >
            <h3>
               Admin Name : 
              Nadya Yuki
            </h3>
            <h3>
               Admin Email : 
              yukiwuki@gmail.com
            </h3>
            <h3>
               Admin Contact : 
              12345678
            </h3>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
